// @rpii/wdio-html-reporter/lib/types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ReportData = exports.InternalReportEvent = exports.TestInfo = exports.SuiteInfo = exports.Metrics = exports.HtmlReporterOptions = void 0;
class HtmlReporterOptions {
    constructor() {
        this.outputDir = "";
        this.filename = "";
        this.templateFilename = "";
        this.templateFuncs = {};
        this.reportTitle = "Please add a Title";
        this.showInBrowser = false;
        this.collapseTests = false;
        this.collapseSuites = false;
        this.useOnAfterCommandForScreenshot = false;
        this.LOG = null;
        this.debug = false;
        this.browserName = "";
        this.removeOutput = true;
    }
}
exports.HtmlReporterOptions = HtmlReporterOptions;
class Metrics {
    constructor() {
        this.passed = 0;
        this.skipped = 0;
        this.failed = 0;
        this.start = "";
        this.end = "";
        this.duration = 0;
    }
}
exports.Metrics = Metrics;
class SuiteInfo {
    constructor(type, suiteStats) {
        this.type = type;
        this.suite = suiteStats;
        this.tests = [];
    }
}
exports.SuiteInfo = SuiteInfo;
class TestInfo {
    constructor(test) {
        this.testStats = test;
        this.events = [];
        this.errorIndex = 0;
    }
}
exports.TestInfo = TestInfo;
class InternalReportEvent {
    constructor(type, value) {
        this.type = type;
        this.value = value;
    }
}
exports.InternalReportEvent = InternalReportEvent;
class ReportData {
    constructor(title, info, suites, metrics, reportFile, browserName) {
        this.info = info;
        this.metrics = metrics;
        this.title = title;
        this.suites = suites;
        this.reportFile = reportFile;
        this.browserName = browserName;
    }
}
exports.ReportData = ReportData;
